api/lcp/collection.py

```diff
@@ -0,0 +1,311 @@
+
+            patron.external_identifier = license['encryption']['user_key']['key_check']
+
+        loan = LoanInfo(
+            licensepool.collection,
+            licensepool.data_source.name,
+            identifier_type=licensepool.identifier.type,
+            identifier=licensepool.identifier.identifier,
+            start_date=today,
+            end_date=expires,
+            fulfillment_info=None,
+            external_identifier=license['id']
+        )
+
+        return loan
+
+    def fulfill(self, patron, pin, licensepool, internal_format=None, part=None, fulfill_part_url=None):
+        """Get the actual resource file to the patron.
+
+        :param patron: A Patron object for the patron who wants to check out the book
+        :type patron: Patron
+
+        :param pin: The patron's alleged password
+        :type pin: string
+
+        :param licensepool: Contains lending info as well as link to parent Identifier
+        :type licensepool: LicensePool
+
+        :param internal_format: A vendor-specific name indicating the format requested by the patron
+        :type internal_format:
+
+        :param part: A vendor-specific identifier indicating that the
+            patron wants to fulfill one specific part of the book
+            (e.g. one chapter of an audiobook), not the whole thing
+        :type part: Any
+
+        :param fulfill_part_url: A function that takes one argument (a
+            vendor-specific part identifier) and returns the URL to use
+            when fulfilling that part
+        :type fulfill_part_url: Any
+
+        :return: a FulfillmentInfo object
+        :rtype: FulfillmentInfo
+        """
+        loan = get_one(self._db, Loan, patron=patron, license_pool=licensepool, on_multiple='interchangeable')
+        license = self._lcp_server.get_license(loan.external_identifier, loan.external_identifier)
+        fulfillment_info = LCPFulfilmentInfo(
+            licensepool.identifier.identifier,
+            licensepool.collection,
+            licensepool.data_source.name,
+            licensepool.identifier.type,
+            content_link=None,
+            content_type=DeliveryMechanism.LCP_DRM,
+            content=license,
+            content_expires=None,
+        )
+
+        return fulfillment_info
+
+    def patron_activity(self, patron, pin):
+        """Returns patron's loans
+
+        :param patron: A Patron object for the patron who wants to check out the book
+        :type patron: Patron
+
+        :param pin: The patron's alleged password
+        :type pin: string
+
+        :return: List of patron's loans
+        :rtype: List[LoanInfo]
+        """
+        # NOTE: We assume that there can be only ONE LCP collection per library
+        loans = self._db\
+            .query(Loan)\
+            .join(LicensePool)\
+            .join(Collection)\
+            .filter(
+                Loan.patron == patron,
+                Collection.id == self._collection_id
+            )
```

Comment:
This will find loans that have expired but have not yet been reaped, or which (not likely to happen) have not started yet. Is that OK?

Reply:
I don't think it's ok. As far as I understand `patron_activity` should return only active loans. What is the best to filter out non-active loans? Should it be something like that?
```python
loans = self._db\
            .query(Loan)\
            .join(LicensePool)\
            .join(Collection)\
            .filter(
                Loan.patron == patron,
                Loan.fulfillment_id != None,
                or_(
                    Loan.start is None,
                    Loan.start >= func.now()
                 ),
                 or_(
                    Loan.end is None,
                    Loan.end < func.now()
                 ),
                Collection.id == self._collection_id
            )
```

Reply:
Yes, that looks good.